tools/webidl_binder.py

```diff
@@ -133,6 +133,15 @@ def build_constructor(name):
 mid_js += build_constructor('WrapperObject')
 
 mid_js += ['''
+/*
+ * For now, the webidl-generated code unconditionally depends on the `assert` function,
+ * but there are certain build modes where emscripten does not define this.
+ * TODO(sbc): Make the usage of assert conditional.
+ */
+if (typeof assert == "undefined") {
+  assert = (cond) => {}
+}
```

Comment:
```suggestion
if (typeof assert === "undefined") {
  var assert = () => {};
}
```

Reply:
This may not work, if the generated code here is in a different scope than where `assert` is defined,
```js
function assert() { .. }
function other() {
  if (typeof assert === "undefined") {
   // assert is always overridden here
    var assert = () => {};
  }
}
```
I think the WebIDL binder output here is usually added in a pre-js, so maybe that would be fine. However, it seems a little risky. Might be simplest to just define a new `wassert` and use that in the WebIDL binder code, perhaps?

Reply:
I think the above example would work just fine wouldn't it `typeof assert === "undefined"` will return false if there is an assert in the output/global scope.

webidl code certainly must be the same scope as the emscripten module code and `--post-js` is the canonical way to use it.

In the long term I hope to add `--debug` flag to webidl_binder.py but for now I think this solution should work fine.

Reply:
Because JS does function hoisting the check will always work regardless of whether this code comes after of before the assert declartion.

For example the following code always prints "good" regardless of whether it is run at the global scope or in a function:

```
  if (typeof assert == 'undefined') {
    console.error('oops');
  } else {
    console.error('good');
  }

  function assert() {
  }
```

Reply:
> I think the above example would work just fine wouldn't it typeof assert === "undefined" will return false if there is an assert in the output/global scope.

No, there is problem with scoping. Here is an extension of my example that is runnable:
```
function assert() {
  throw "true assert";
}

function other() {
  if (typeof assert === "undefined") {
   // assert is always overridden here
    var assert = () => {};
  }
  assert(1 == 0);
}

other();
```
Running that does *not* throw `true assert`, showing the assert was always overridden.

`var` is scoped to the function, so entering `other` means that the `assert` in the outer scope does not matter. Just the presence of the `var` means that `assert` is equal to `undefined` in that scope (as it is undefined until it actually reaches the `var`).